File: MasterCatalog.Domain/Models/Category.cs
namespace MasterCatalog.Domain.Models
{
    public class Category
    {
        public int? CategoryID { get; set; }

        public string CategoryName { get; set; }

        public string CategoryDescription { get; set; }

        public int? ParentCategoryID { get; set; }

        public Category? ParentCategory { get; set; }

        public List<Category>? SubCategories { get; set; }
    }
}

File: MasterCatalog.Web/wwwroot/js/models/category.js
export class Category {
    categoryID;
    categoryName;
    categoryNameError;
    categoryDescription;
    categoryDescriptionError;
    subCategories = [];
    parentCategoryID;
    parentCategory = {};

    constructor(id, name, description, parent, subCategories) {
        this.categoryID = id;
        this.categoryName = name;
        this.categoryDescription = description;
        this.parentCategory = parent;
        this.subCategories = subCategories;

        if (parent) {
            this.parentCategoryID = parent.categoryID;
        }
    }

    validate() {
        this.clearErrors();

        let valid = true;
        const name = this.categoryName;
        if (!name) {
            this.categoryNameError = 'Category name is Required.';
            valid = false;
        }

        const description = this.categoryDescription;
        if (!description) {
            this.categoryDescriptionError = 'Category description is Required.';
            valid = false;
        }

        return valid;
    }

    clearErrors() {
        this.categoryNameError = null;
        this.categoryDescriptionError = null;
    }
}